# 실무예제 5-2
## 다음은 2015년 국내대학현황이다. 재적학생수를 엔트로피-기반, 카이제곱-결합, 직관적분할 방식을 각각 활용하여 20개 구간 내외로 이산화하시오. 
### 데이터 파일 : ch5-2(국내대학현황).csv
### 원본 투플수 : 1,720개

In [1]:
# ch5-2.py
import pandas as pd
import numpy as np

# 데이터로드 (ch5-2(국내대학현황).csv : 데이터 원본 파일)
# encoding : 윈도우즈 환경에서의 한글 처리
# engine : python 3.6에서 한글이 포함된 파일이름 사용
rawData = pd.read_csv('.jupyter/ch5-2(국내대학현황).csv', encoding='CP949', engine='python')

In [20]:
rawData

학제                    학교명  지역     설립  재적학생수
0        전문대학                 한국철도대학  경기     국립    180
1        전문대학                 강원도립대학  강원     공립   1658
2        전문대학               경남도립거창대학  경남     공립   1614
3        전문대학               경남도립남해대학  경남     공립   1639
4        전문대학                경북도립대학교  경북     공립   1389
5        전문대학                 충북도립대학  충북     공립   1815
6        전문대학                충남도립대학교  충남     공립   1747
7        전문대학                한국복지대학교  경기     국립    735
8        전문대학               가톨릭상지대학교  경북     사립   2859
9        전문대학                군산간호대학교  전북     사립   1241
10       전문대학                  거제대학교  경남     사립   2525
11       전문대학                경남정보대학교  부산     사립  11748
12       전문대학                  호산대학교  경북     사립   2832
13       전문대학                  경민대학교  경기     사립   8154
14       전문대학                대구미래대학교  경북     사립   1430
15       전문대학                영남외국어대학  경북     사립   1070
16       전문대학                경북전문대학교  경북     사립   4290
17       전문대학                  경복대학교  경기     사립   8672
18       전문대학                경인여자대학교  인천     사립   5309
19       전문대학                 서라벌대학교  경북     사립   1270
20       전문대학                계명문화대학교  대구     사립   8149
21       전문대학                계원예술대학교  경기     사립   4361
22       전문대학                광양보건대학교  전남     사립   2439
23       전문대학                광주보건대학교  광주     사립   5376
24       전문대학                  구미대학교  경북     사립   8306
25       전문대학                   서해대학  전북     사립   2084
26       전문대학                  군장대학교  전북     사립   3028
27       전문대학                  강동대학교  충북     사립   7079
28       전문대학                 고구려대학교  전남     사립   2364
29       전문대학                기독간호대학교  광주     사립    713
...       ...                    ...  ..    ...    ...
1690  대학부설대학원          진주산업대학교 산업대학원  경남     국립      7
1691  대학부설대학원        진주산업대학교 벤처창업대학원  경남     국립      4
1692  대학부설대학원           한국교통대학교산업대학원  충북     국립      2
1693  대학부설대학원      한국교통대학교경영행정외국어대학원  충북     국립      3
1694  대학부설대학원    한국산업기술대학교 산업기술경영대학원  경기     사립      1
1695  대학부설대학원        한국방송통신대학교 경영대학원  서울     국립    234
1696  대학부설대학원          한국방송통신대학교 대학원  서울     국립   2388
1697  대학부설대학원             초당대학교산업대학원  전남     사립     14
1698  대학부설대학원        청운대학교 산업기술경영대학원  충남     사립    115
1699  대학부설대학원          경운대학교 산업정보대학원  경북     사립      1
1700  대학부설대학원       경희사이버대학교 호텔관광대학원  서울     사립    114
1701  대학부설대학원       경희사이버대학교 문화창조대학원  서울     사립    212
1702  대학부설대학원       대구사이버대학교 휴먼케어대학원  경북     사립    115
1703  대학부설대학원      부산디지털대학교 휴먼서비스대학원  부산     사립     88
1704  대학부설대학원       서울사이버대학교 상담심리대학원  서울     사립    236
1705  대학부설대학원      서울사이버대학교 휴먼서비스대학원  서울     사립    169
1706  대학부설대학원          세종사이버대학교  대학원  서울     사립    132
1707  대학부설대학원       원광디지털대학교 웰빙문화대학원  전북     사립    132
1708  대학부설대학원       고려사이버대학교 융합정보대학원  서울     사립     57
1709  대학부설대학원         한양사이버대학교 경영대학원  서울     사립    326
1710  대학부설대학원       한양사이버대학교 교육정보대학원  서울     사립     82
1711  대학부설대학원        한양사이버대학교 디자인대학원  서울     사립     46
1712  대학부설대학원        한양사이버대학교 부동산대학원  서울     사립     70
1713  대학부설대학원      한양사이버대학교 휴먼서비스대학원  서울     사립    284
1714  대학부설대학원  사이버한국외국어대학교 TESOL 대학원  서울     사립    137
1715  대학부설대학원          한국과학기술원 일반대학원  대전  특별법법인   6962
1716  대학부설대학원          광주과학기술원 일반대학원  광주  특별법법인   1167
1717  대학부설대학원        대구경북과학기술원 일반대학원  대구  특별법법인    321
1718  대학부설대학원        한국전통문화대학교 일반대학원  충남  특별법국립     35
1719  대학부설대학원    한국전통문화대학교 문화유산융합대학원  충남  특별법국립     76

[1720 rows x 5 columns]

In [6]:
# 재적학생수 속성을 벡터(리스트)로 저장
v_재적학생수 = list(rawData.loc[:, '재적학생수'])

# 오름차순 정렬
v_재적학생수.sort()

### 재적학생수 기준 이산화를 위해서 재적학생수 백터(리스트)를 정렬해 놓음

In [22]:
v_재적학생수

[1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 3,
 3,
 3,
 3,
 3,
 3,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 5,
 6,
 6,
 6,
 6,
 6,
 7,
 7,
 7,
 7,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 10,
 10,
 10,
 10,
 10,
 11,
 11,
 11,
 11,
 11,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 13,
 13,
 13,
 13,
 13,
 13,
 14,
 14,
 14,
 14,
 14,
 14,
 15,
 15,
 15,
 15,
 16,
 16,
 16,
 16,
 17,
 17,
 17,
 17,
 17,
 17,
 17,
 18,
 19,
 20,
 20,
 20,
 20,
 20,
 20,
 20,
 20,
 20,
 20,
 21,
 21,
 21,
 21,
 22,
 22,
 22,
 22,
 22,
 22,
 22,
 22,
 22,
 22,
 22,
 22,
 23,
 23,
 23,
 23,
 24,
 24,
 24,
 24,
 24,
 24,
 25,
 25,
 25,
 25,
 25,
 26,
 26,
 26,
 26,
 26,
 26,
 27,
 27,
 27,
 27,
 27,
 27,
 27,
 28,
 28,
 28,
 28,
 28,
 28,
 29,
 29,
 29,
 29,
 30,
 30,
 30,
 30,
 31,
 31,
 31,
 31,
 31,
 31,
 31,
 31,
 31,
 31,
 32,
 32,
 32,
 32,
 32,
 32,
 32,
 32,
 32,
 32,
 33,
 33,
 33,
 33,
 33,
 34,

## 1) 엔트로피(entropy)-기반 재적학생수 이산화

In [7]:
# 1) 엔트로피(entropy)-기반 재적학생수 이산화
# l_bound: 최소점, u_bound: 최대점, s_point: 분할점
# v_s_points: 구간경계점을 저정하는 벡터
l_bound = 0
u_bound = len(v_재적학생수) - 1
v_s_points = []  # 벡터(리스트) 초기화

# 엔트로피 계산을 위한 scipy.stats 모듈 임포트
from scipy.stats import entropy

# 기대정보요구량을 구하는 함수 정의
# 해당 구간에서 최소기대정보요구량을 나타내는 지점을 찾고 그 점을 경계로 재귀적으로 호출하는 구조
def info_gain(l_bound, u_bound) :
    i = l_bound + 1

    # 최소기대정보요구량의 위치 찾기
    while (i < u_bound) :
        # 기대정보요구량 계산
        v_info_gain = entropy(v_재적학생수[l_bound:(i + 1)])*(i - l_bound + 1) \
                      / (u_bound - l_bound + 1) \
                      + entropy(v_재적학생수[(i + 1):(u_bound + 1)]) * (u_bound - i) \
                      / (u_bound - l_bound + 1)

        # 최소값 찾기
        if (i == l_bound + 1) :
            min_info_gain = v_info_gain
            min_idx = i
        else :
            if (v_info_gain < min_info_gain) :
                min_info_gain = v_info_gain
                min_idx = i

        i = i + 1

    # 구간 수 20개 내외로 분할하기 위해 임계값을 4.0로 설정(trial and error 방식으로 찾음)
    if (min_info_gain < 4.0) :
        return 1

    # 최소지점 위치를 v_s_points 벡터에 추가
    v_s_points.extend([min_idx])
    
    # 경계점을 중심으로 한 재귀적 함수 호출
    info_gain(l_bound, min_idx)
    info_gain(min_idx, u_bound)

# 함수 호출
info_gain(l_bound, u_bound)

# 경계점 벡터 정렬(오름차순)
v_s_points.sort()

# 리스트 변수에 엔트로피 이산화 결과값(재적학생수, 대학수)을 행 단위로 저장
list_of_disc = []
for i in range(0, len(v_s_points)) :
    idx = v_s_points[i]
    list_of_disc.append([v_재적학생수[idx], idx])

# 데이터프레임 생성
rawData_disc = pd.DataFrame(list_of_disc, columns=['재적학생수', '대학수'])

### 기대정보요구량 계산을 바탕으로 구간 경계점을 재귀적으로 찾아나가기 위하여 info_gain()이라는 사용자 함수 설계
#### 최소 기대정보요구량의 위치를 찾기 위해서 해당 구간 내 모든 지점의 기대정보요구량을 비교
### entropy 임계치는 문제에서 제시한 구간 수인 20개 내외를 맞추기 위해 시행착오(trial & error) 방식으로 탐색 

In [8]:
v_s_points

[109,
 201,
 289,
 378,
 465,
 554,
 641,
 729,
 815,
 902,
 988,
 1075,
 1129,
 1184,
 1239,
 1297,
 1401,
 1507,
 1609]

In [9]:
list_of_disc

[[16, 109],
 [32, 201],
 [45, 289],
 [60, 378],
 [73, 465],
 [92, 554],
 [110, 641],
 [134, 729],
 [164, 815],
 [197, 902],
 [252, 988],
 [326, 1075],
 [410, 1129],
 [519, 1184],
 [674, 1239],
 [1070, 1297],
 [2195, 1401],
 [5195, 1507],
 [10106, 1609]]

In [10]:
rawData_disc

재적학생수   대학수
0      16   109
1      32   201
2      45   289
3      60   378
4      73   465
5      92   554
6     110   641
7     134   729
8     164   815
9     197   902
10    252   988
11    326  1075
12    410  1129
13    519  1184
14    674  1239
15   1070  1297
16   2195  1401
17   5195  1507
18  10106  1609

### 위는 엔트로피 이산화에 의해 구간화된 구간경계점을 출력한 결과임
#### 총 20개의 구간이 생성됨
### 엔트로피 기반 이산화의 결과를 담은 데이터프레임 : rawData_disc
#### 첫 번째 열 : 재적학생수
#### 두 번째 열 : 대학수
#### 예컨대, 재적학생수가 16명까지인 대학수가 110개(0~109는 110개), 17~32명까지인 대학수가 92개(201-109는 92개)임. 재적학생수가 10,106명 초과인 대학 수는 전체대학 수가 1,720개이므로 111개로 볼 수 있음

## 2) 카이제곱-결합 이산화

In [4]:
# 2) 카이제곱-결합 이산화
# 재적학생수와 지역 간의 카이제곱 검정에 의한 재적학생수 이산화

# 원본 데이터파일에서 '시군'과 '지정구분' 속성만 추출
rawData_kai = rawData.loc[:, ('지역','재적학생수')]

import cx_Oracle      # Oracle DB 연동을 위한 cx_Oracle 패키지 임포트
from scipy import stats  # 카이제곱검정을 위한 scipy 패키지 중 stats 모듈 임포트

# Oracle DB 연결
# 접속정보(connection string) : ID/PASS@CONNECTION_ALIAS
# CONNECTION_ALIAS : Oracle TNSNAMES.ORA 파일에 있는 접속정보 별칭(ALIAS)
conn_ora = cx_Oracle.connect("prep1/prep1@XE")

# DB 커서(Cursor) 선언
cur = conn_ora.cursor()

src_table = "d_base5_2"  # 사용할 Oracle 소스 테이블명 지정
ord_table = "d_base_ord"  # 소스테이블에서 재적학생수 기준으로 정렬하여 순번을 매긴 테이블명 지정
cont_table = "d_base_cont"  # 사용할 분할표(contingency table)을 저장할 Oracle 소스 테이블명 지정

# 데이터프레임(rawData)에 저장된 데이터를 Oracle 테이블(d_base5_2)에 입력하기 위한 로직
# d_base5_2 테이블 존재하는지 체크하는 함수
def table_exists(name=None, con=None):
    sql = "select table_name from user_tables where table_name='MYTABLE'".replace('MYTABLE', name.upper())
    df = pd.read_sql(sql, con)

    # 테이블이 존재하면 True, 그렇지 않으면 False 반환
    exists = True if len(df) > 0 else False
    return exists

# 테이블(d_base5_2) 생성 (테이블이 이미 존재한다면 TRUNCATE TABLE)
if table_exists(src_table, conn_ora):
    cur.execute("TRUNCATE TABLE " + src_table)
else:
    cur.execute("create table " + src_table + " ( \
               지역 varchar2(10), \
               재적학생수 number(6))")

# Sequence 구조를 Dictionary 구조((element, value))로 변환하는 함수
# 예: ("Matt", 1) -> {'1':'Matt', '2':1}
# INSERT INTO ... VALUES (:1, :2, ...) 에서 바인드 변수값을 주기위해 Dictionary item 구조 사용
def convertSequenceToDict(list):
    dict = {}
    argList = range(1, len(list) + 1)
    for k, v in zip(argList, list):
        dict[str(k)] = v
    return dict

# 데이터프레임에 저장된 데이터를 Oracle 테이블로 입력(insert)
cols = [k for k in rawData_kai.dtypes.index]
colnames = ','.join(cols)
colpos = ', '.join([':' + str(i + 1) for i, f in enumerate(cols)])
insert_sql = 'INSERT INTO %s (%s) VALUES (%s)' % (src_table, colnames, colpos)

# INSERT INTO ... VALUES (:1, :2, ...)의 바인드 변수 값을 저장하는 Dictionary 구조 생성
data = [convertSequenceToDict(rec) for rec in rawData_kai.values]

# 바인드 변수와 Dictionary 데이터구조를 활용하여 Bulk Insertion 구현
cur.executemany(insert_sql, data)
conn_ora.commit()

# 재적학생수 기준으로 오름차순 정렬
# 순번 테이블(d_base_ord) 생성 (테이블이 이미 존재한다면 TRUNCATE TABLE)
if table_exists(ord_table, conn_ora):
    cur.execute("TRUNCATE TABLE " + ord_table)
else:
    cur.execute("CREATE TABLE " + ord_table + " ( \
               지역 varchar2(10), \
               재적학생수 number(6), \
               순번 number(4))")

cur.execute("INSERT INTO " + ord_table +
            " select 지역, 재적학생수, \
            row_number() over (order by 재적학생수) - 1 순번 from " + src_table)

conn_ora.commit()

# df_재적학생수 : 재적학생수의 오름차순으로 정렬된 재적학생수 데이터프레임
df_재적학생수 = pd.read_sql("select 재적학생수 from " + ord_table + " order by 재적학생수", con=conn_ora)

# 분할표 테이블(d_base_cont) 생성 (테이블이 이미 존재한다면 TRUNCATE TABLE)
if table_exists(cont_table, conn_ora):
    cur.execute("TRUNCATE TABLE " + cont_table)
else:
    cur.execute("CREATE TABLE " + cont_table + " ( \
               지역 varchar2(10), \
               구간 varchar2(7))")

# 카이제곱 검정을 통해 구간병합을 할지 말지를 결정하는 함수
def chi_process(l_bound1, u_bound1, l_bound2, u_bound2) :

    # 이전 구간의 분할표 데이터 삭제
    cur.execute("TRUNCATE TABLE " + cont_table)

    # 관측도수 분할표 생성
    cur.execute("INSERT INTO " + cont_table +
                " select 지역 \
                , case when 순번 between " + str(l_bound1) + " and " + str(u_bound1) +
                " then '구간1' else '구간2' end 구간 \
                from " + ord_table +
                " where 순번 between " + str(l_bound1) + " and " + str(u_bound2))

    conn_ora.commit()

    # 관측도수/기대도수 분할표 생성 (1차원 배열 형식)
    count_df = pd.read_sql("select a.지역 \
                          , a.구간 \
                          , nvl(b.관측도수,0) 관측도수 /*관측되지 않은 (지역+구간)은 0으로 처리*/ \
                          , a.기대도수_지역 * a.기대도수_구간 / a.기대도수_전체 기대도수 \
                          from ( \
                                 select x.지역, y.구간 \
                                      , x.기대도수_지역 \
                                      , y.기대도수_구간 \
                                      , x.기대도수_전체 \
                                 from ( select 지역 \
                                           , count(*) 기대도수_지역 /* 지역 속성의 cardinality */ \
                                           , sum(count(*)) over () 기대도수_전체 /*전체 행 개수*/ \
                                        from " + cont_table + " \
                                        group by 지역 ) x, \
                                      ( select 구간 \
                                           , count(*) 기대도수_구간 /* 구간 속성의 cardinality */ \
                                        from " + cont_table + " \
                                        group by 구간 ) y ) a, \
                                 ( select 지역 \
                                        , 구간 \
                                        , count(*) 관측도수 /* 지역, 구간 별 실제 행 개수 */ \
                                   from " + cont_table + " \
                                   group by 지역, 구간 ) b \
                          where a.지역 = b.지역(+) /* 특정 (지역+구간) 값은 존재하지 않을 수 \
                                                    있어서 외부조인으로 처리 */ \
                                and a.구간 = b.구간(+)", con=conn_ora)

    # 2개 속성에 대한 자유도(degree of freedom) 갭 구하기
    # A 속성에 대한 cardinality = a, B 속성에 대한 cardinality = b라 가정
    # cardinality : 서로 다른 속성값의 개수
    # 분할표 전체 행 갯수(a*b) 구하기
    tot_rows = count_df.shape[0]

    # A 속성의 cardianlity(a)와 B 속성의 cardinality(b) 구하기
    count_df2 = pd.read_sql("select count(distinct 지역) 도수_지역 \
                               , count(distinct 구간) 도수_구간 \
                          from " + cont_table, con=conn_ora)

    # cardinality 갭 [(a*b-1) - (a-1)*(b-1)] 구하기
    v_ddof = (tot_rows - 1) - (count_df2['도수_지역'] - 1) * (count_df2['도수_구간'] - 1)

    obs_array = count_df['관측도수']  # obs_array : 관측도수를 저장하는 1차원 배열
    exp_array = count_df['기대도수']  # exp_array : 기대도수를 저장하는 1차원 배열

    # stats.chisquare() : 카이제곱검정 함수
    chis = stats.chisquare(obs_array, exp_array, ddof=v_ddof)

    # pvalue 임계치(0.35)는 구간수 20개 내외를 맞추기 위해 시행착오(trial & error) 방식으로 찾음
    if (chis.pvalue[0] > 0.35) :
        return(True)    # 구간들이 클래스와 독립적이므로 구간병합
    else :
        return(False)   # 구간들이 클래스와 독립적이라 볼 수 없으므로 병합하지 않음

# gugan_len : 초기 구간 단위를 30으로 일정하게 잡음
# boundary_idx : 구간 경계점의 위치를 저장하는 벡터
gugan_len = 30
i = 0
boundary_idx = []

# 구간 경계점의 위치를 지정 (30단위로)
while (i*gugan_len <= len(df_재적학생수)) :
    boundary_idx.extend([i*gugan_len])
    i = i + 1

# 마지막 구간 경계점 지점 (반드시 30단위로 된다고 볼 수 없으므로 별도 지정)
boundary_idx.extend([len(df_재적학생수)-1])

# 인접 구간과의 카이제곱 검정을 통해 p-value가 특정 임계치 이상인 경우(병합이 필요한 경우)는
# 해당 경계점의 boundary_idx 값을 -1로 세팅
# boundary_idx를 차례로 순환하면서 병합지점을 찾고,
# 순환 동안 병합이 한번도 발생하지 않을 시에 while루프 탈출
while (1) :
    merge_cnt = 0
    i = 0
    ind = 1

    # l_bound1: 인접 두 구간 중 첫번째 구간의 왼쪽 경계점
    # u_bound1: 인접 두 구간 중 첫번째 구간의 오른쪽 경계점
    # l_bound2: 인접 두 구간 중 두번째 구간의 왼쪽 경계점
    # u_bound2: 인접 두 구간 중 두번째 구간의 오른쪽 경계점
    while (i < len(boundary_idx)-1) :
        if (boundary_idx[i] == -1) :
            i = i + 1
        elif (ind == 1) :
            l_bound1 = boundary_idx[i]
            i = i + 1
            ind = ind + 1
        elif (ind == 2) :
            u_bound1 = boundary_idx[i] - 1
            l_bound2 = boundary_idx[i]
            merge_idx = i
            i = i + 1
            ind = ind + 1
        elif (ind == 3) :
            u_bound2 = boundary_idx[i] - 1

            # 인접 구간의 경계점을 패러미터로 주고 카이제곱 검정 사용자정의 함수 호출
            if (chi_process(l_bound1, u_bound1, l_bound2, u_bound2)) :
                # 병합이 필요한 경우에는 두 구간 경계점의 boundary_idx 값을 -1로 세팅
                boundary_idx[merge_idx] = -1
                merge_cnt = merge_cnt + 1  # 병합 카운트 1증가
            ind = 1

    # 병합 카운트가 0일 때(순환주기 동안 병합이 한번도 일어나지 않을 때), 루프 탈출
    if (merge_cnt == 0) :
        break

cur.close()  # 커서(cursor) 종료
conn_ora.close()  # Oracle connection 종료

# 결과보기
i = 1
v_s_points = []   # 구간 분할 경계점을 저장하는 벡터 초기화

# boundary_idx가 -1일 경우는 구간 병합을 의미하므로 분할 경계점에서 제외
while (i < len(boundary_idx)-1) :
    if (boundary_idx[i] != -1) :
        v_s_points.extend([boundary_idx[i] - 1])
    i = i + 1

# 리스트 변수에 엔트로피 이산화 결과값(재적학생수, 대학수)을 행 단위로 저장
list_of_disc = []
for i in range(0, len(v_s_points)) :
    idx = v_s_points[i]
    list_of_disc.append([df_재적학생수.loc[idx,'재적학생수'], idx])

# 데이터프레임 생성
rawData_disc = pd.DataFrame(list_of_disc, columns=['재적학생수', '구간경계점(대학수)'])

### 카이제곱-결합 이산화는 bottom-up 방식이기 때문에 엔트로피-기반 이산화 기법처럼 재귀적 방식을 적용하기는 힘듬
#### 구간을 병합하는 방식이기 때문에 초기 구간을 설정하는 것이 필요함
#### 본 예제에서는 초기 구간을 30단위의 일정 길이로 설정함
### 알고리즘 개요
#### 구간을 처음부터 끝까지 순회하면서 인접 두 구간을 대상으로 카이제곱 검정을 수행
#### 카이제곱 검정 결과, 유의수준(pvalue)이 일정 임계치 이상이면, 클래스(여기서는 지역)와 독립적으로 판단하여 해당 두 구간을 병합
#### 한 순회주기 동안 한 번의 병합도 일어나지 않으면 구간병합 프로세스 종료
### chi_process() : 카이제곱 검정을 수행하는 사용자정의 함수
#### 패러미터로 들어온 2개의 구간(구간1: [l_bound1, u_bound1], 구간2: [l_bound1, u-bound2]) 간의 카이제곱 검정을 수행해서 구간을 병합할지 여부를 결정
#### pvalue 임계치는 문제에서 제시한 구간 수인 20개 내외를 맞추기 위해 시행착오(trial & error) 방식으로 탐색하여 구하였다. 

In [5]:
rawData_disc

재적학생수  구간경계점(대학수)
0      28         179
1      37         239
2      43         269
3      66         419
4     101         599
5     113         659
6     123         689
7     132         719
8     207         929
9     230         959
10    252         989
11    274        1019
12    301        1049
13    483        1169
14   1231        1319
15   5064        1499
16   6052        1529
17   9230        1589
18  10297        1619
19  12462        1649
20  26523        1709

### 위는 카이제곱-결합 이산화에 의해 구간화된 구간경계점을 출력한 결과임
#### 총 22개의 구간이 생성됨
### 카이제곱-결합 이산화의 결과를 담은 데이터프레임 : rawData_disc
#### 첫 번째 열 : 재적학생수
#### 두 번째 열 : 대학수
#### 재적학생수 기준으로 정렬된 결과이며, 180번(0~179번)째 대학(재적학생수 28명 이하)까지가 첫 번째 구간이고 181번째에서 240번째 대학까지가 두 번째 구간임. 마지막 구간은 1711번째 대학(재적학생수 26,523명 초과)부터임
#### 병합된 구간의  대학수가 210개까지 존재

## 3) 직관적 분할에 의한 재적학생수 이산화

In [17]:
# 3) 직관적 분할에 의한 재적학생수 이산화
# v_s_points : 재적학생수 구간 경계점을 저장하는 벡터(리스트)
v_s_points = []  # 초기화

# 3-4-5 규칙을 재귀적으로 적용하는 함수
# l_bound : 구간 최소값
# u_bound : 구간 최대값
# v_depth : 함수 호출 depth
def intuitive_split(l_bound, u_bound, v_depth) :

    # 문자형 변환
    l_bound_str = str(l_bound)
    u_bound_str = str(u_bound)

    # 문자길이 반환
    ms_pos = len(u_bound_str)

    # 서로 다른 가장 큰 유효숫자(Most Significant Bit, MSB) 의 갯수 구하기
    i = l_bound
    ms_v = ""

    # 구간 내 숫자별 MSB를 구하여 ms_v에 저장
    while (i <= u_bound) :
        ele_str = str(i)
        ms_v = ms_v + ele_str[0]
        i = i+1

    # 중복 배제
    ms_v_unique = set(ms_v)

    # ms_card : 서로 다른 MSB의 갯수 저장
    ms_card = len(ms_v_unique)

    # ms_card의 수에 따라 분할 구간 수를 다르게 함 (3-4-5 규칙에 따라)
    if (ms_card == 3 or ms_card == 6 or ms_card == 9) :

        # gugan_len : 단위 구간 길이
        gugan_len = round((u_bound - l_bound) / 3, 0)

        # 2번째 중첩 호출(v_depth: 2)에서 구간 분할을 끝냄
        if (v_depth == 2) :
            # 구간 경계점 벡터 v_s_points에 요소 추가
            v_s_points.extend([l_bound])
            v_s_points.extend([l_bound + gugan_len])
            v_s_points.extend([l_bound + gugan_len * 2])
            v_s_points.extend([u_bound])
            return (True)
        else :
            # 1번째 함수 호출에서는 구간 범위를 달리하여 재귀적 함수 호출
            intuitive_split(l_bound, l_bound + gugan_len, v_depth + 1)
            intuitive_split(l_bound + gugan_len, l_bound + gugan_len * 2, v_depth + 1)
            intuitive_split(l_bound + gugan_len * 2, u_bound, v_depth + 1)
    elif (ms_card == 7) :
        gugan_len = round((u_bound - l_bound) / 7, 0)
        if (v_depth == 2) :
            v_s_points.extend([l_bound])
            v_s_points.extend([l_bound + gugan_len * 2])
            v_s_points.extend([l_bound + gugan_len * 5])
            v_s_points.extend([u_bound])
            return (True)
        else :
            intuitive_split(l_bound, l_bound + gugan_len * 2, v_depth + 1)
            intuitive_split(l_bound + gugan_len * 2, l_bound + gugan_len * 5, v_depth + 1)
            intuitive_split(l_bound + gugan_len * 5, u_bound, v_depth + 1)
    elif (ms_card == 2 or ms_card == 4 or ms_card == 8) :
        gugan_len = round((u_bound - l_bound) / 4, 0)
        if (v_depth == 2) :
            v_s_points.extend([l_bound])
            v_s_points.extend([l_bound + gugan_len])
            v_s_points.extend([l_bound + gugan_len * 2])
            v_s_points.extend([l_bound + gugan_len * 3])
            v_s_points.extend([u_bound])
            return (True)
        else :
            intuitive_split(l_bound, l_bound + gugan_len, v_depth + 1)
            intuitive_split(l_bound + gugan_len, l_bound + gugan_len * 2, v_depth + 1)
            intuitive_split(l_bound + gugan_len * 2, l_bound + gugan_len * 3, v_depth + 1)
            intuitive_split(l_bound + gugan_len * 3, u_bound, v_depth + 1)
    elif (ms_card == 1 or ms_card == 5 or ms_card == 10) :
        gugan_len = round((u_bound - l_bound) / 5, 0)
        if (v_depth == 2) :
            v_s_points.extend([l_bound])
            v_s_points.extend([l_bound + gugan_len])
            v_s_points.extend([l_bound + gugan_len * 2])
            v_s_points.extend([l_bound + gugan_len * 3])
            v_s_points.extend([l_bound + gugan_len * 4])
            v_s_points.extend([u_bound])
            return (True)
        else :
            intuitive_split(l_bound, l_bound + gugan_len, v_depth + 1)
            intuitive_split(l_bound + gugan_len, l_bound + gugan_len * 2, v_depth + 1)
            intuitive_split(l_bound + gugan_len * 2, l_bound + gugan_len * 3, v_depth + 1)
            intuitive_split(l_bound + gugan_len * 3, l_bound + gugan_len * 4, v_depth + 1)
            intuitive_split(l_bound + gugan_len * 4, u_bound, v_depth + 1)

# 함수 호출
v_재적학생수.sort()
intuitive_split(v_재적학생수[0], v_재적학생수[len(v_재적학생수)-1], 1)

# 중복값 제거 및 오름차순 정렬
list_of_disc = list(set(v_s_points))
list_of_disc.sort()

# 데이터프레임 생성
rawData_disc = pd.DataFrame(list_of_disc, columns=['재적학생수'])

### intuitive_split() : 3-4-5 규칙에 따라 구간의 직관적 분할을 수행하는 사용자정의 함수이다. 
#### 이 함수의 재귀적으로 호출로 구간을 분할해 나감
#### 함수 호출 깊이(depth)를 2로 제한함. 이는 구간의 수의 관련이 있는데, 결과적으로 11개의 구간이 생성된다.
##### depth를 3으로 할 경우는 구간의 수가 50개로 늘어나 너무 많이 구간이 쪼개어 짐
### 결과를 데이터프레임 : rawData_disc
#### 재적학생수의 구간 경계점을 저장하고 있음 

In [18]:
rawData_disc

재적학생수
0       1.0
1    3814.0
2    7627.0
3   11440.0
4   14300.0
5   17160.0
6   20020.0
7   22879.0
8   25739.0
9   28599.0
10  31459.0
11  34319.0

### 위 결과화면은 [1, 3814], [3815, 7627], ... [31460, 34319] 등 11개의 구간을 보여줌
### 구간 길이를 수학적으로 분할한 관계로 앞의 두 기법과는 달리, 실제 구간경계점에 해당하는 재적학생수를 갖는 대학이 없을 수 있음
### 데이터 분포와 상관없이 구간의 길이는 비교적 일정함 